### ISSUE TITLE
lensless example

### ISSUE THREAD

**Issue by fatmalama** [FPtje/miso-component-example#1]
(I've been trying for a whole week to achieve the functionality of zoom with mtl/transformers and couldn't find any)

**Comment by FPtje** [comment]
Are you trying to get the same functionality in the `Effect` monad or are you trying to recurse into a component, but without using `zoom`?

**Comment by fatmalama** [comment]
recurse into a component

(I can replace the model (state) in a Transition without lens setters with Control.Monad.State.modify, but I haven't found a way to zoom (getting the component's state wrapped inside the model of the parent's Transition, running the child's updateModel and wrapping it back))

**Comment by FPtje** [comment]
Sorry for taking a while. This is a bit of a tricky one, and I hadn't had time to give you a proper example. I have to work this out myself, so I'm going to write down my thinking steps here:

Let's look at the snippet that updates the right button in the components example:
```haskell
updateModel :: Action -> Transition Action Model ()
updateModel action = case action of
  [...]
  RightButtonAction act -> do
          zoom mRightButton $ Button.updateModel iRightButton act
  [...]
```

Let's look at what's happening here. `Transition` is defined to be a strict `StateT model (Writer [Sub action])` (in Miso's source `frontend-src/Miso/Types.hs`). We're gonna have to mess with `StateT`, so let's get some imports. 

```Haskell
import qualified Control.Monad.Trans.State.Strict as State
import Control.Monad.Trans.Class ( lift )
```
I'll explain the use of `lift` later.

Let's look at some types. Our `Main`'s `updateModel` function has return type `Transition Action Model ()`, and the `Button`'s `updateModel` has return type `Transition action Button.Model ()`. Let's substitute for the definition of `Transition`.


| Location | Original return type  | Substituted|
| ------------- | ------------- |------------- |
| Main | `Transition Main.Action Main.Model ()`  | `State.StateT Main.Model (Writer [Sub Main.Action])`  |
| Button | `Transition action Button.Model ()`  | `State.StateT Button.Model (Writer [Sub action])` |

Important is that the Button's `updateModel` has `action` as a type variable. The `Model` however, is specific to the Button. We'll need to convert that somehow. Staring at Hoogle, I found that `State.execStateT` actually does a lot of what we need: You give it (in our case) the Button's `StateT`, the Button's model, and it'll reward you with an updated Button model. Sadly, it won't work as is:

- Its first argument requires the Button's model. Luckily, we can get it through `State.get` and running `_mRightButton` on it.
- It returns its result in the "parent" `Writer` monad. Luckily, the Button's `updateModel` has `action` as a type variable. That means we can run it in `Main`'s `updateModel` using `lift`.

Once we have an updated `Button.Model`, we can update the `Main`'s `StateT` with `State.put`. 

The end result is as follows:


```haskell
RightButtonAction act -> do
    m <- State.get
    rightButton <-
      lift $ State.execStateT (Button.updateModel iRightButton act) $ _mRightButton m
    State.put $ m { _mRightButton = rightButton }
```

Does this answer your question?

**Comment by fatmalama** [comment]
yes!!!
I also [asked on stack overflow](https://stackoverflow.com/questions/54583848/haskell-mtl-transformers-equivalent-of-lens-zooming-a-state) and got a different answer (but also working) by Snoyman the yesod guy. I actually did try lift originally but I don't know how I screwed up.

as for this as a bug (rather feature request) - I think lens is too big of a library and I spent like a week trying to reason how zoom worked to avoid lens. maybe add the equivalence as a comment? e.g.

```
RightButtonAction act -> do
  zoom mRightButton $ Button.updateModel iRightButton act
  -- If not using the lens library:
  --m <- State.get
  --rightButton <-
  --  lift $ State.execStateT (Button.updateModel iRightButton act) $ _mRightButton m
  --State.put $ m { _mRightButton = rightButton }
```
